File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/model/IDOTakingMedicineReminderParamModel.kt
//
//  IDOTakingMedicineReminderParamModel.kt
//  protocol_channel
//
//  Created by hedongyang on 2023/10/30.
//

package com.idosmart.model

import com.google.gson.*
import java.lang.reflect.Type

///
/// Set Taking Medicine Reminder Event Code
open class IDOTakingMedicineReminderParamModel(
    takingMedicineId: Int,
    onOff: Int,
    startHour: Int,
    startMinute: Int,
    endHour: Int,
    endMinute: Int,
    repeats: HashSet<IDOWeek>,
    interval: Int,
    doNotDisturbOnOff: Int,
    doNotDisturbStartHour: Int,
    doNotDisturbStartMinute: Int,
    doNotDisturbEndHour: Int,
    doNotDisturbEndMinute: Int
) : IDOBaseAdapterModel<IDOTakingMedicineReminderParamModel> {

    /// ID ranging from 1 to 5
    var takingMedicineId: Int = takingMedicineId

    /// 0 for off
    /// 1 for on
    var onOff: Int = onOff

    /// Starting hour of the reminder
    var startHour: Int = startHour

    /// Starting minute of the reminder
    var startMinute: Int = startMinute

    /// Ending hour of the reminder
    var endHour: Int = endHour

    /// Ending minute of the reminder
    var endMinute: Int = endMinute

    /// Repeat
    var repeats: HashSet<IDOWeek> = repeats

    /// Reminder interval in minutes
    /// Default is 60 minutes
    var interval: Int = interval

    /// Do not disturb time period switch
    /// 0 for off
    /// 1 for on
    /// Default is off
    var doNotDisturbOnOff: Int = doNotDisturbOnOff

    /// Do not disturb start hour
    var doNotDisturbStartHour: Int = doNotDisturbStartHour

    /// Do not disturb start minute
    var doNotDisturbStartMinute: Int = doNotDisturbStartMinute

    /// Do not disturb end hour
    var doNotDisturbEndHour: Int = doNotDisturbEndHour

    /// Do not disturb end minute
    var doNotDisturbEndMinute: Int = doNotDisturbEndMinute
    override fun getDeSerializer(): IDOBaseModelDeSerializer<IDOTakingMedicineReminderParamModel>? {
        return TakingMedicineReminderParamModelSerializer()
    }

    override fun toString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


internal class TakingMedicineReminderParamModelSerializer : IDOBaseModelDeSerializer<IDOTakingMedicineReminderParamModel> {
    override fun serialize(src: IDOTakingMedicineReminderParamModel, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val jsonObject = JsonObject()
        jsonObject.addProperty("taking_medicine_id", src.takingMedicineId)
        jsonObject.addProperty("on_off", src.onOff)
        jsonObject.addProperty("start_hour", src.startHour)
        jsonObject.addProperty("start_minute", src.startMinute)
        jsonObject.addProperty("end_hour", src.endHour)
        jsonObject.addProperty("end_minute", src.endMinute)
        jsonObject.addProperty("interval", src.interval)
        jsonObject.addProperty("do_not_disturb_on_off", src.doNotDisturbOnOff)
        jsonObject.addProperty("do_not_disturb_start_hour", src.doNotDisturbStartHour)
        jsonObject.addProperty("do_not_disturb_start_minute", src.doNotDisturbStartMinute)
        jsonObject.addProperty("do_not_disturb_end_hour", src.doNotDisturbEndHour)
        jsonObject.addProperty("do_not_disturb_end_minute", src.doNotDisturbEndMinute)

        var num = 0
        src.repeats.forEach { week ->
            num = num or (1 shl week.rawValue + 1)
        }
        jsonObject.addProperty("repeat", num)
        return jsonObject
    }

    override fun deserialize(
        json: JsonElement?,
        typeOfT: Type?,
        context: JsonDeserializationContext?
    ): IDOTakingMedicineReminderParamModel? {
        json?.let {
            val jsonObject = it.asJsonObject
            val takingMedicineId = jsonObject.get("taking_medicine_id").asInt
            val onOff = jsonObject.get("on_off").asInt
            val startHour = jsonObject.get("start_hour").asInt
            val startMinute = jsonObject.get("start_minute").asInt
            val endHour = jsonObject.get("end_hour").asInt
            val endMinute = jsonObject.get("end_minute").asInt
            val interval = jsonObject.get("interval").asInt
            val doNotDisturbOnOff = jsonObject.get("do_not_disturb_on_off").asInt
            val doNotDisturbStartHour = jsonObject.get("do_not_disturb_start_hour").asInt
            val doNotDisturbStartMinute = jsonObject.get("do_not_disturb_start_minute").asInt
            val doNotDisturbEndHour = jsonObject.get("do_not_disturb_end_hour").asInt
            val doNotDisturbEndMinute = jsonObject.get("do_not_disturb_end_minute").asInt

            val repeat = jsonObject.get("repeat").asInt
            val items = HashSet<IDOWeek>()
            for (i in 0..6) {
                if (repeat and (1 shl (i + 1)) != 0) {
                    items.add(IDOWeek.fromRawValue(i))
                }
            }

            return IDOTakingMedicineReminderParamModel(
                takingMedicineId,
                onOff,
                startHour,
                startMinute,
                endHour,
                endMinute,
                items,
                interval,
                doNotDisturbOnOff,
                doNotDisturbStartHour,
                doNotDisturbStartMinute,
                doNotDisturbEndHour,
                doNotDisturbEndMinute
            )
        }
        return null
    }
}
    

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/Model/IDOTakingMedicineReminderParamModel.swift
//
//  IDOTakingMedicineReminderParamModel.swift
//  protocol_channel
//
//  Created by hc on 2023/10/12.
//

import Foundation
/// Set Taking Medicine Reminder Event Code
public class IDOTakingMedicineReminderParamModel: NSObject, IDOBaseModel {
    /// ID ranging from 1 to 5
    public var takingMedicineId: Int
    /// 0 for off
    /// 1 for on
    public var onOff: Int
    /// Starting hour of the reminder
    public var startHour: Int
    /// Starting minute of the reminder
    public var startMinute: Int
    /// Ending hour of the reminder
    public var endHour: Int
    /// Ending minute of the reminder
    public var endMinute: Int
    /// Repeat
    public var repeats: Set<IDOWeek>
    /// Reminder interval in minutes
    /// Default is 60 minutes
    public var interval: Int
    /// Do not disturb time period switch
    /// 0 for off
    /// 1 for on
    /// Default is off
    public var doNotDisturbOnOff: Int
    /// Do not disturb start hour
    public var doNotDisturbStartHour: Int
    /// Do not disturb start minute
    public var doNotDisturbStartMinute: Int
    /// Do not disturb end hour
    public var doNotDisturbEndHour: Int
    /// Do not disturb end minute
    public var doNotDisturbEndMinute: Int

    enum CodingKeys: String, CodingKey {
        case takingMedicineId = "taking_medicine_id"
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case repeats = "repeat"
        case interval = "interval"
        case doNotDisturbOnOff = "do_not_disturb_on_off"
        case doNotDisturbStartHour = "do_not_disturb_start_hour"
        case doNotDisturbStartMinute = "do_not_disturb_start_minute"
        case doNotDisturbEndHour = "do_not_disturb_end_hour"
        case doNotDisturbEndMinute = "do_not_disturb_end_minute"
    }

    public init(takingMedicineId: Int, onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, repeats: Set<IDOWeek>, interval: Int, doNotDisturbOnOff: Int, doNotDisturbStartHour: Int, doNotDisturbStartMinute: Int, doNotDisturbEndHour: Int, doNotDisturbEndMinute: Int) {
        self.takingMedicineId = takingMedicineId
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.repeats = repeats
        self.interval = interval
        self.doNotDisturbOnOff = doNotDisturbOnOff
        self.doNotDisturbStartHour = doNotDisturbStartHour
        self.doNotDisturbStartMinute = doNotDisturbStartMinute
        self.doNotDisturbEndHour = doNotDisturbEndHour
        self.doNotDisturbEndMinute = doNotDisturbEndMinute
    }

    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        takingMedicineId = try container.decode(Int.self, forKey: .takingMedicineId)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        interval = try container.decode(Int.self, forKey: .interval)
        doNotDisturbOnOff = try container.decode(Int.self, forKey: .doNotDisturbOnOff)
        doNotDisturbStartHour = try container.decode(Int.self, forKey: .doNotDisturbStartHour)
        doNotDisturbStartMinute = try container.decode(Int.self, forKey: .doNotDisturbStartMinute)
        doNotDisturbEndHour = try container.decode(Int.self, forKey: .doNotDisturbEndHour)
        doNotDisturbEndMinute = try container.decode(Int.self, forKey: .doNotDisturbEndMinute)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = Set<IDOWeek>()
            for i in 0...6 {
                if value & (1 << (i + 1)) != 0 {
                    items.insert(IDOWeek(rawValue: i)!)
                }
            }
            repeats = items
        } else {
            repeats = Set<IDOWeek>()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(takingMedicineId, forKey: .takingMedicineId)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(interval, forKey: .interval)
        try container.encode(doNotDisturbOnOff, forKey: .doNotDisturbOnOff)
        try container.encode(doNotDisturbStartHour, forKey: .doNotDisturbStartHour)
        try container.encode(doNotDisturbStartMinute, forKey: .doNotDisturbStartMinute)
        try container.encode(doNotDisturbEndHour, forKey: .doNotDisturbEndHour)
        try container.encode(doNotDisturbEndMinute, forKey: .doNotDisturbEndMinute)

        var value = 0
        for item in repeats {
            value |= (1 << (item.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }
    
    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}


/// Set Taking Medicine Reminder Event Code
@objcMembers
public class IDOTakingMedicineReminderParamModelObjc: NSObject, IDOBaseModel {
    /// ID ranging from 1 to 5
    public var takingMedicineId: Int
    /// 0 for off
    /// 1 for on
    public var onOff: Int
    /// Starting hour of the reminder
    public var startHour: Int
    /// Starting minute of the reminder
    public var startMinute: Int
    /// Ending hour of the reminder
    public var endHour: Int
    /// Ending minute of the reminder
    public var endMinute: Int
    /// Repeat
    public var repeats: [IDOWeekObjc]
    /// Reminder interval in minutes
    /// Default is 60 minutes
    public var interval: Int
    /// Do not disturb time period switch
    /// 0 for off
    /// 1 for on
    /// Default is off
    public var doNotDisturbOnOff: Int
    /// Do not disturb start hour
    public var doNotDisturbStartHour: Int
    /// Do not disturb start minute
    public var doNotDisturbStartMinute: Int
    /// Do not disturb end hour
    public var doNotDisturbEndHour: Int
    /// Do not disturb end minute
    public var doNotDisturbEndMinute: Int

    enum CodingKeys: String, CodingKey {
        case takingMedicineId = "taking_medicine_id"
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case repeats = "repeat"
        case interval = "interval"
        case doNotDisturbOnOff = "do_not_disturb_on_off"
        case doNotDisturbStartHour = "do_not_disturb_start_hour"
        case doNotDisturbStartMinute = "do_not_disturb_start_minute"
        case doNotDisturbEndHour = "do_not_disturb_end_hour"
        case doNotDisturbEndMinute = "do_not_disturb_end_minute"
    }

    public init(takingMedicineId: Int, onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, repeats: [IDOWeekObjc], interval: Int, doNotDisturbOnOff: Int, doNotDisturbStartHour: Int, doNotDisturbStartMinute: Int, doNotDisturbEndHour: Int, doNotDisturbEndMinute: Int) {
        self.takingMedicineId = takingMedicineId
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.repeats = repeats
        self.interval = interval
        self.doNotDisturbOnOff = doNotDisturbOnOff
        self.doNotDisturbStartHour = doNotDisturbStartHour
        self.doNotDisturbStartMinute = doNotDisturbStartMinute
        self.doNotDisturbEndHour = doNotDisturbEndHour
        self.doNotDisturbEndMinute = doNotDisturbEndMinute
    }

    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        takingMedicineId = try container.decode(Int.self, forKey: .takingMedicineId)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        interval = try container.decode(Int.self, forKey: .interval)
        doNotDisturbOnOff = try container.decode(Int.self, forKey: .doNotDisturbOnOff)
        doNotDisturbStartHour = try container.decode(Int.self, forKey: .doNotDisturbStartHour)
        doNotDisturbStartMinute = try container.decode(Int.self, forKey: .doNotDisturbStartMinute)
        doNotDisturbEndHour = try container.decode(Int.self, forKey: .doNotDisturbEndHour)
        doNotDisturbEndMinute = try container.decode(Int.self, forKey: .doNotDisturbEndMinute)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = [IDOWeekObjc]()
            for i in 0...6 {
                if value & (1 << (i + 1)) != 0 {
                    items.append(IDOWeekObjc(week: IDOWeek(rawValue: i)!))
                }
            }
            repeats = items
        } else {
            repeats = [IDOWeekObjc]()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(takingMedicineId, forKey: .takingMedicineId)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(interval, forKey: .interval)
        try container.encode(doNotDisturbOnOff, forKey: .doNotDisturbOnOff)
        try container.encode(doNotDisturbStartHour, forKey: .doNotDisturbStartHour)
        try container.encode(doNotDisturbStartMinute, forKey: .doNotDisturbStartMinute)
        try container.encode(doNotDisturbEndHour, forKey: .doNotDisturbEndHour)
        try container.encode(doNotDisturbEndMinute, forKey: .doNotDisturbEndMinute)

        var value = 0
        for item in repeats {
            value |= (1 << (item.week.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }
    
    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}
